text/0005-is-operator.md

```diff
@@ -0,0 +1,176 @@
+
+With an `is` operator, users could instead write the following.
+
+```
+permit(
+  principal is User,
+  action == Action::"view",
+  resource is File in Folder::"public"
+);
+```
+
+`is` acts as a dynamic check on the runtime types of `principal` and `resource`.
+This puts the important type information into the policy itself, as opposed to a separate schema, making it easy to tell what the policy does.
+
+## Use case 2: Reduce `has` checks
+
+The `is` operator can also be used in some places where `has` is currently required.
+For example, consider the following policy taken from the Cedar [document cloud example](https://github.com/cedar-policy/cedar-examples/tree/main/cedar-example-use-cases/document_cloud).
+
+```
+forbid (principal, action, resource)
+when
+{
+  resource has owner &&
+  principal != resource.owner &&
+  resource has isPrivate &&
+  resource.isPrivate
+};
+```
+
+This policy acts on any type of resource, but can only be satisfied when the resource has `owner` and `isPrivate` attributes.
+If the resource does not have these attributes, then the policy is a no-op.
+Either way, we need to check whether the attributes exist using `has` to avoid runtime errors.
+
+Now say that only one entity type (e.g., `File`) has the fields `owner` and `isPrivate`.
+Then this policy could be written more succinctly as follows.
+
+```
+forbid (principal, action, resource is File)
+when
+{
+  principal != resource.owner &&
+  resource.isPrivate
+};
+```
+
+# Detailed design
+
+## Changes required to parser/AST
+
+- Update the grammar and AST to support expressions of the form `e is et` where `et` is a `Name` (e.g., `Namespace::EntityType`).
+- Update the grammar and AST to support expressions of the form `e1 is et in e2` (but not `e1 in e2 is et`, because this is more difficult to mentally parse).
+- Update the grammar to allow expressions like `e is et` and `e1 is et in e2` in the policy scope.
+
+## Changes required to evaluator
+
+`e is et` will evaluate to true iff `et` is a `Name` (e.g., `Namespace::EntityType`) and `e` is an expression that is an entity of type `et`.
+Otherwise it will evaluate to false.
+To maintain the property that "conditions in the scope never error," `e is et` will never return an error; it will simply return false if any error occurs.
+
+Namespaces are considered "part of" the entity type. So:
+
+- `User::"alice" is User` is true
+- `Namespace::User::"alice" is User` is false
+- `Namespace::User::"alice" is Namespace::User` is true
+- `User::"alice" is Namespace::User` is false
+
+## Changes required to validator
+
+`e is et` will be typed as `True` if the validator can guarantee that expression `e` is an entity of type `et`, or `False` if the validator can guarantee that `e` is a non-entity type or an entity not of type `et`.
+Otherwise, the validator gives `e is et` type `Bool`.
+
+To improve typing precision, we can extend the current notion of "effects" to track information about entity types.
```

Comment:
This sounds doable. It would be a necessary enhancement to make template slots useful if we accept RFC #3.